src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder_ListPatterns.cs

```diff
@@ -55,17 +56,19 @@ private Tests MakeTestsAndBindingsForListPattern(BoundDagTemp input, BoundListPa
 
                         if (slice.Pattern is BoundPattern slicePattern)
                         {
-                            Debug.Assert(slice.IndexerAccess is not null);
-                            Debug.Assert(index <= 0);
-                            Debug.Assert(slice.ReceiverPlaceholder is not null);
-                            Debug.Assert(slice.ArgumentPlaceholder is not null);
+                            // We faced a slice pattern with a declaration pattern of the same narrowed type inside, e.g. `[<pattern>, ..var between, <pattern>]`
+                            // This pattern always evaluates to true, but causes slicing operation and even may allocate (e.g. when slicing an array or string)
+                            // Thus it makes sense to hold it and "evaluate" after all other patterns in the list matched
+                            if (slicePattern.Kind == BoundKind.DeclarationPattern && slicePattern.InputType.Equals(slicePattern.NarrowedType, TypeCompareKind.AllNullableIgnoreOptions))
```

Comment:
Is this check really necessary? the narrowed type for the slice won't escape from within the element itself. Then you can inline the local function outside the loop.

Reply:
> The narrowed type check is needed to prevent order of evaluation change when types are different since in such case we cannot guarantee that the pattern always evaluates to `true`

Matching the slice pattern last will change the ordering (of the `Slice` call) anyways? but that is obviously not a requirement - allowing this optimization. I don't think narrowed type of the slice pattern is at issue here because it won't affect any other patterns e.g. the remaining subpatterns in the list. 
> we cannot tell, whether it is cheaper to perform slicing pattern check or check of the last N patterns in the list

As you mentioned, `Slice` may actually allocate for non-span types so it's worth considering making it the last test to perform regardless of what subpattern we have. As for overall ordering of the tests based on cost, there's https://github.com/dotnet/roslyn/issues/29033.

Unless this complicates subsumption checking I think it should be ok to this uniformly. I may have missed something from previous discussions, but @jcouv can weigh in on weather or not that's an appropriate change to make.

Reply:
I'd be okay to generalize the optimization.